README

Welcome to balance.fm!

This is Last.fm's fork of the balance tool, offering some neat new
features (like monitoring support) and bunch of bugfixes.

See debian/changelog for a list of changes or just pull the repo from

  https://github.com/lastfm/balance.fm

and use git log to see what has changed in even more detail.

Find the original README of balance below.

Enjoy,
Marcus Holland-Moritz <marcus@last.fm>

--8<------

Welcome to the source code distribution of balance !

Installation:
-------------

Linux (i386/itanium), FreeBSD, BSD/OS:
        make
        make install

Solaris:
        Uncomment the LIBRARY and INSTALL definitions in the Makefile.
        make
        make install

If you have any problems just change the very simple Makefile to
fit your needs. Shared memory should be enabled on your machine.

For cygwin uncomment the cygwin definitions in the Makefile first.

Information about new releases, mailing lists and other related issues
can be found at the balance homepage http://balance.sourceforge.net .

Reports of successful use of balance are appreciated.

Have fun...
Thomas Obermair <obermair@acm.org>


debian/changelog
balance.fm (1.0.1) unstable; urgency=low

  * Update repo link in manpage
  * Transfer repository to lastfm organization
  * Fix manpage installation issues

 -- Marcus Holland-Moritz <marcus@last.fm>  Wed, 04 Jul 2012 10:22:25 +0000

balance.fm (1.0.0) unstable; urgency=low

  * Initial release, based on balance 3.54
  * Add TCP keepalive configuration support
  * Fix and refactor logging/reporting code
  * Fix some off-by-one errors
  * Fix some inconsistent exit codes
  * Improve signal handling and shared memory handling
  * Improve shutdown behaviour
  * Added monitoring support
  * Documentation overhaul

 -- Marcus Holland-Moritz <marcus@last.fm>  Mon, 13 Feb 2012 15:53:02 +0000


balance.h
/*
 * balance.h
 */

/*
 * $Id: balance.h,v 1.1 2010/01/29 10:40:16 t Exp $
 */

#include <stdio.h>
#include <stdarg.h>
#include <stdbool.h>
#include <strings.h>
#include <ctype.h>
#include <errno.h>
#include <netdb.h>
#include <signal.h>
#include <stdlib.h>
#include <sysexits.h>
#include <syslog.h>
#include <poll.h>
#ifndef NO_MMAP
#include <unistd.h>
#include <sys/mman.h>
#endif

#ifdef __FreeBSD__
#define BalanceBSD 1
#endif

#ifdef bsdi
#define BalanceBSD 1
#endif

#ifdef BSD
#define BalanceBSD 1
#endif

#ifdef MAC_OS_X_VERSION_10_4
#define BalanceBSD 1
#endif

#ifndef BalanceBSD
#include <sys/resource.h>
#endif

#include <sys/socket.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <sys/shm.h>

#include <netinet/in.h>
#include <arpa/inet.h>
#include <netinet/tcp.h>        /* for TCP_NODELAY definition */

#if defined(TCP_KEEPIDLE) && defined(TCP_KEEPINTVL) && defined(TCP_KEEPCNT)
# define BALANCE_CAN_KEEPALIVE 1
#else
# define BALANCE_CAN_KEEPALIVE 0
#endif

#ifndef __GNUC__
# define __attribute__(x)  /*NOTHING*/
#endif

/* solaris 9, solaris 10 do not have INADDR_NONE */
#ifndef INADDR_NONE
#define INADDR_NONE ((unsigned long) -1)
#endif

#define MAXTXSIZE       (32*1024)
#define FILENAMELEN     1024
/*
 * this should be a directory that isn't cleaned up periodically, or at
 * reboot of the machine (/tmp is cleaned at reboot on many OS versions)
 */
#define SHMDIR          "/var/run/balance/"
#define SHMFILESUFFIX   ".shm"

#define MAXCHANNELS             64      /* max channels in group          */
#define MAXGROUPS               16      /* max groups                     */
#define MAXINPUTLINE            128     /* max line in input mode         */
#define DEFAULTTIMEOUT          5       /* timeout for unreachable hosts  */
#define DEFAULTSELTIMEOUT       0       /* timeout for select             */
#define DEFAULT_MON_INTERVAL    60      /* 1 minute monitoring interval   */

enum channel_status {
  CS_DISABLED,                  /* hard disabled (manually in shell)  */
  CS_ENABLED,                   /* enabled                            */
  CS_DISABLED_SOFT              /* soft disabled (by monitoring)      */
};

typedef struct {
  int time;
  int intvl;
  int probes;
} KEEPALIVE;

typedef struct {
  enum channel_status status;
  int port;
  struct in_addr ipaddr;
  int c;                        /* current # of connections           */
  int tc;                       /* total # of connections             */
  int maxc;                     /* max # of connections, 0 = no limit */
  unsigned long long bsent;     /* bytes sent                         */
  unsigned long long breceived; /* bytes received                     */
} CHANNEL;

#define GROUP_RR        0       /* Round Robin            */
#define GROUP_HASH      1       /* Hash on Client Address */

typedef struct {
  int nchannels;                /* number of channels in group         */
  int current;                  /* current channel in group            */
  int type;                     /* either GROUP_RR or GROUP_HASH       */
  CHANNEL channels[MAXCHANNELS];
} GROUP;

typedef struct {
  int   release;
  int   subrelease;
  int   patchlevel;
  int   pid;
  int   monitor_pid;
  bool  monitor_enabled;
  int   ngroups;
  GROUP groups[MAXGROUPS];
} COMMON;

/*
 * Macros to access various elements of struct GROUP and struct CHANNEL
 * within COMMON array
 *
 * a       pointer to variable of type COMMON
 * g       group index
 * i       channel index
 */

#define cmn_group(a,g)           ((a)->groups[(g)])
#define grp_nchannels(a,g)       (cmn_group((a),(g)).nchannels)
#define grp_current(a,g)         (cmn_group((a),(g)).current)
#define grp_type(a,g)            (cmn_group((a),(g)).type)
#define grp_channel(a,g,i)       (cmn_group((a),(g)).channels[(i)])
#define chn_status(a,g,i)        (grp_channel((a),(g),(i)).status)
#define chn_port(a,g,i)          (grp_channel((a),(g),(i)).port)
#define chn_ipaddr(a,g,i)        (grp_channel((a),(g),(i)).ipaddr)
#define chn_c(a,g,i)             (grp_channel((a),(g),(i)).c)
#define chn_tc(a,g,i)            (grp_channel((a),(g),(i)).tc)
#define chn_maxc(a,g,i)          (grp_channel((a),(g),(i)).maxc)
#define chn_bsent(a,g,i)         (grp_channel((a),(g),(i)).bsent)
#define chn_breceived(a,g,i)     (grp_channel((a),(g),(i)).breceived)

enum monitor_action_type {
  MA_CONNECT,                   /* performs a TCP connect to host:port */
  MA_COMMAND                    /* run an arbitrary command and checks */
                                /*   its exit code (0 = passed)        */
  /* MA_PING  might be useful, but can for now be emulated using command=/bin/ping... */
};

enum monitor_status {
  MS_UNKNOWN,                   /* no tests have been performed yet    */
  MS_PASSED,                    /* all tests have passed so far        */
  MS_FAILED,                    /* at least one test has failed        */
  MS_ERROR                      /* an error occurred during a test     */
};

struct monitor_defaults {
  float connect_timeout;
};

struct monitor_action {
  enum monitor_action_type type;
  struct monitor_action *next;
  union {
    struct {
      float timeout;
    } connect;
    struct {
      char *cmdline;
      size_t num_pass;
      int *pass;
    } command;
  } u;
};

struct monitor_spec {
  struct monitor_action *action_list;
  int interval;                  /* monitoring interval in seconds     */
  bool enable;                   /* enable channels when tests pass    */
  bool disable;                  /* disable channels when tests fail   */
};

struct monitor_info {
  int grp;
  int cha;
  enum monitor_status status;
};

/*
 * function prototypes
 */
unsigned int hash_fold(const void *, int);
ssize_t writen(int, const unsigned char *, size_t);
int err_dump(const char *text);
struct monitor_spec *monitor_spec_parse(const char *str, const struct monitor_defaults *defaults);
void monitor_spec_dump(FILE *fh, const struct monitor_spec *spec);
void monitor_spec_free(struct monitor_spec *spec);
char *monitor_command_format(const char *format, const char *host, int port);


butils.c
/*
 * Copyright (c) 2000-2004,2005 by Inlab Software GmbH, Gruenwald, Germany.
 * All rights reserved.
 *
 */

/*
 * $Id: butils.c,v 1.1 2010/01/29 10:40:16 t Exp $
 */

#include "balance.h"

unsigned int hash_fold(const void *ptr, int len)
{
  const unsigned char *s = ptr;
  unsigned int rc = 0;

  for (int i = 0; i < len; i++) {
    rc = s[i] + 31 * rc;
  }

  return rc;
}

ssize_t writen(int fd, const unsigned char *ptr, size_t nbytes)
{
  int nleft;
  ssize_t nwritten;

  nleft = nbytes;

  while (nleft > 0) {
    nwritten = write(fd, ptr, nleft);
    if (nwritten <= 0) {
      return nwritten;        /* error */
    }
    nleft -= nwritten;
    ptr += nwritten;
  }

  return nbytes - nleft;
}

static void *safe_malloc(size_t size)
{
  void *m = malloc(size);
  if (m == NULL) {
    err_dump("out of memory");
  }
  return m;
}

static void *safe_realloc(void *ptr, size_t size)
{
  void *m = realloc(ptr, size);
  if (size > 0 && m == NULL) {
    err_dump("out of memory");
  }
  return m;
}

static char *get_token(const char **str, const char *delim_list)
{
  const char *p = *str;
  char *token = safe_malloc(strlen(p) + 1);
  char *t = token;

  while (*p) {
    const char *d = delim_list;

    while (*p && *p == '\\') {
      if (*++p) {
        switch (*p) {
          case 'a': *t++ = '\a'; break;
          case 'b': *t++ = '\b'; break;
          case 't': *t++ = '\t'; break;
          case 'n': *t++ = '\n'; break;
          case 'v': *t++ = '\v'; break;
          case 'f': *t++ = '\f'; break;
          case 'r': *t++ = '\r'; break;
          default:  *t++ = *p;   break;
        }

        p++;
      }
    }

    while (*p && *d && *p != *d) {
      d++;
    }

    if (*p == '\0' || *p == *d) {
      break;
    }

    *t++ = *p++;
  }

  *str = p;
  *t++ = '\0';

  return safe_realloc(token, t - token);
}

static struct monitor_action *append_action(struct monitor_action ***pppma, enum monitor_action_type type)
{
  struct monitor_action *ma = safe_malloc(sizeof(struct monitor_action));
  ma->type = type;
  ma->next = **pppma;
  **pppma = ma;
  *pppma = &ma->next;
  return ma;
}

struct monitor_spec *monitor_spec_parse(const char *str, const struct monitor_defaults *defaults)
{
  struct monitor_spec *ms = safe_malloc(sizeof(struct monitor_spec));

  ms->action_list = NULL;
  ms->interval = DEFAULT_MON_INTERVAL;
  ms->enable = true;
  ms->disable = true;

  struct monitor_action **ppma = &ms->action_list;
  struct monitor_action *cur_ma = NULL;

  while (*str) {
    char *k = get_token(&str, ":=");
    char *v = NULL;

    if (*str == '=') {
      str++;
      v = get_token(&str, ":");
    }

    if (strcmp(k, "connect") == 0) {
      if (v) {
        err_dump("unexpected value for 'connect'");
      }
      cur_ma = append_action(&ppma, MA_CONNECT);
      cur_ma->u.connect.timeout = defaults->connect_timeout;
    } else if (strcmp(k, "command") == 0) {
      if (v == NULL) {
        err_dump("'command' action requires a value");
      }
      cur_ma = append_action(&ppma, MA_COMMAND);
      cur_ma->u.command.cmdline = v;
      cur_ma->u.command.num_pass = 1;
      cur_ma->u.command.pass = safe_malloc(sizeof(int));
      cur_ma->u.command.pass[0] = 0;
    } else if (strcmp(k, "interval") == 0) {
      if (v == NULL) {
        err_dump("'interval' option requires a value");
      }
      char *end;
      ms->interval = strtol(v, &end, 0);
      if (*end || ms->interval <= 0) {
        err_dump("error parsing interval");
      }
      free(v);
    } else if (strcmp(k, "timeout") == 0) {
      if (v == NULL) {
        err_dump("'timeout' option requires a value");
      }
      if (cur_ma == NULL) {
        err_dump("'timeout' option given without action");
      }
      if (cur_ma->type != MA_CONNECT) {
        err_dump("'timeout' option unsupported for this action");
      }
      char *end;
      cur_ma->u.connect.timeout = strtod(v, &end);
      if (*end) {
        err_dump("error parsing timeout");
      }
      if (cur_ma->u.connect.timeout < 1e-6f) {
        err_dump("timeout outside of supported range");
      }
      free(v);
    } else if (strcmp(k, "pass") == 0) {
      if (v == NULL) {
        err_dump("'pass' option requires a value");
      }
      if (cur_ma == NULL) {
        err_dump("'pass' option given without action");
      }
      if (cur_ma->type != MA_COMMAND) {
        err_dump("'pass' option unsupported for this action");
      }
      cur_ma->u.command.num_pass = 1;
      for (const char *c = v; *c; c++) {
        if (*c == ',') {
          cur_ma->u.command.num_pass++;
        }
      }
      cur_ma->u.command.pass = safe_realloc(cur_ma->u.command.pass,
                                            cur_ma->u.command.num_pass*sizeof(int));
      char *end = v;
      for (size_t i = 0; i < cur_ma->u.command.num_pass; i++) {
        cur_ma->u.command.pass[i] = strtol(end, &end, 0);
        if (*end != (i < cur_ma->u.command.num_pass - 1 ? ',' : '\0')) {
          err_dump("error parsing pass value list");
        }
        if (*end) {
          end++;
        }
      }
      free(v);
    } else if (strcmp(k, "noenable") == 0) {
      if (v) {
        err_dump("unexpected value for 'noenable'");
      }
      ms->enable = false;
    } else if (strcmp(k, "nodisable") == 0) {
      if (v) {
        err_dump("unexpected value for 'nodisable'");
      }
      ms->disable = false;
    } else {
      err_dump("invalid monitoring spec");
    }

    free(k);

    if (*str) {
      str++;
    }
  }

  return ms;
}

void monitor_spec_free(struct monitor_spec *spec)
{
  if (spec)
  {
    struct monitor_action *ma = spec->action_list;

    while (ma) {
      switch (ma->type) {
        case MA_COMMAND:
          free(ma->u.command.cmdline);
          free(ma->u.command.pass);
          break;
        default:
          break;
      }
      struct monitor_action *d = ma;
      ma = ma->next;
      free(d);
    }

    free(spec);
  }
}

void monitor_spec_dump(FILE *fh, const struct monitor_spec *spec)
{
  if (spec)
  {
    fprintf(fh, "monitoring setup:\n");
    fprintf(fh, "  check interval: %d seconds\n", spec->interval);
    fprintf(fh, "  channels will %sbe automatically enabled\n", spec->enable ? "" : "not ");
    fprintf(fh, "  channels will %sbe automatically disabled\n", spec->disable ? "" : "not ");

    const struct monitor_action *ma = spec->action_list;

    fprintf(fh, "  checks:%s\n", ma ? "" : " none");

    while (ma) {
      switch (ma->type) {
        case MA_CONNECT:
          fprintf(fh, "    - connect to channel host [timeout: %.2fs]\n", ma->u.connect.timeout);
          break;
        case MA_COMMAND:
          fprintf(fh, "    - run command: %s [pass exit code%s: ",
                      ma->u.command.cmdline, ma->u.command.num_pass > 1 ? "s" : "");
          for (size_t i = 0; i < ma->u.command.num_pass; i++) {
            fprintf(fh, "%s%d", i == 0 ? "" : ", ", ma->u.command.pass[i]);
          }
          fprintf(fh, "]\n");
          break;
        default:
          break;
      }

      ma = ma->next;
    }
  }
}

static int command_format(char *buffer, int buflen, const char *format, const char *host, int port)
{
  int len = 0;
  char portbuf[32];
  sprintf(portbuf, "%d", port);

#define APPEND_C(c)                            \
        do {                                   \
          if (buffer && len < buflen) {        \
            buffer[len] = (c);                 \
          }                                    \
          len++;                               \
        } while (0)

#define APPEND_S(s)                            \
        do {                                   \
          const char *scp = (s);               \
          int sl = strlen(scp);                \
          if (buffer && len < buflen) {        \
            int cp = buflen - len;             \
            if (sl < cp) cp = sl;              \
            strncpy(buffer + len, (scp), cp);  \
          }                                    \
          len += sl;                           \
        } while (0)

  for (const char *f = format; *f; f++) {
    if (*f == '%' && f[1]) {
      switch (*++f) {
        case 'H':
          APPEND_S(host);
          break;
        case 'P':
          APPEND_S(portbuf);
          break;
        case '%':
           APPEND_C('%');
           break;
        default:
           APPEND_C('%');
           APPEND_C(*f);
           break;
      }
    } else {
      APPEND_C(*f);
    }
  }

  APPEND_C('\0');

#undef APPEND_C
#undef APPEND_S

  return len;
}

char *monitor_command_format(const char *format, const char *host, int port)
{
  int len = command_format(NULL, 0, format, host, port);
  char *buf = safe_malloc(len);
  command_format(buf, len, format, host, port);
  return buf;
}
